common/lockmanager_test.go

```diff
@@ -0,0 +1,355 @@
+
+    Lock(key1);
+
+    threadFinishBlock.Add(1);
+    key2LockedBlock.Add(1);
+    key1UnlockBlock.Add(1);
+
+    go func() {
+        defer threadFinishBlock.Done();
+
+        // Lock key2 while key1 is already locked.
+        Lock(key2);
+
+        key2LockedBlock.Done();
+
+        // Wait for the main thread to unlock key1.
+        key1UnlockBlock.Wait();
+
+        err := Unlock(key2);
+        if (err != nil) {
+            test.Fatalf("Failed to unlock inside the thread.");
+        }
+    }()
+
+    // Wait for the thread to lock key2 before unlocking key1.
+    key2LockedBlock.Wait();
+
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+
+    key1UnlockBlock.Done();
+
+    // Wait for the thread to finish unlocking key2.
+    threadFinishBlock.Wait();
+}
+
+// Lock(1) -> %Lock(2) -> %Unlock(2) -> Unlock(1).
+// % denotes a thread.
+func TestThread2UnlockFirst(test *testing.T) {
+    Clear();
+    defer Clear();
+
+    key1 := "testkey1";
+    key2 := "testkey2";
+    var key2LockUnlockBlock sync.WaitGroup;
+
+    Lock(key1);
+
+    key2LockUnlockBlock.Add(1);
+
+    go func() {
+        defer key2LockUnlockBlock.Done();
+
+        // Lock key2 while key1 is already locked.
+        Lock(key2);
+
+        err := Unlock(key2);
+        if (err != nil) {
+            test.Fatalf("Failed to unlock inside the thread.");
+        }
+    }()
+
+    // Wait for the thread to lock and unlock key2 before unlocking key1.
+    key2LockUnlockBlock.Wait();
+
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+}
+
+// Lock(1) -> %Lock(1) -> Unlock(1) -> %Unlock(1).
+// % denotes a thread.
+func TestLockingTwiceWithSameKeyConcurrently(test *testing.T) {
+    Clear();
+    defer Clear();
+
+    key1 := "testkey1";
+    var threadLockBlock sync.WaitGroup;
+    var threadFinishBlock sync.WaitGroup;
+
+    threadLockBlock.Add(1);
+    threadFinishBlock.Add(1);
+
+    // Lock for the first time.
+    Lock(key1);
+
+    // This thread tries to lock key1 a second time but is blocked
+    // until key1 gets unlocked for the first time.
+    go func() {
+        defer threadFinishBlock.Done();
+
+        threadLockBlock.Done();
+
+        Lock(key1);
+        err := Unlock(key1);
+        if (err != nil) {
+            test.Fatalf("Failed to unlock for the second time.");
+        }
+    }()
+
+    // Wait for the second thread to try to lock key1
+    // while the lock is being used by the first thread.
+    threadLockBlock.Wait();
+    
+    // Small sleep to ensure the thread tries to lock key1.
+    time.Sleep(10 * time.Millisecond);
+
+    // Unlock for the first time.
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock for the first time.");
+    }
+    
+    // Wait for the thread to lock and unlock key1.
+    threadFinishBlock.Wait();
+}
+
+func TestStaleLockWithNonStaleLock(test *testing.T) {
```

Comment:
I don't like the test name, it's not very clear.
What about something like "TestStaleLockRemovalWithNonStaleLock" or "TestLockRetentionWithNonStaleLock".

This comment applies to several of your tests.
Review the test names.